File: src/kitchen_sync/adapters/ssh.py
"""SSH/rsync environment adapter."""

from __future__ import annotations

import subprocess
from pathlib import Path

from kitchen_sync.adapters.base import Environment


class SshEnvironment(Environment):
    """Adapter that deploys via SSH and rsync."""

    def __init__(self, host: str, targets: dict | None = None):
        super().__init__(targets)
        self.host = host

    def is_available(self) -> bool:
        result = subprocess.run(
            ["ssh", "-o", "ConnectTimeout=5", self.host, "true"],
            capture_output=True,
        )
        return result.returncode == 0

    def run(self, cmd: list[str], **kwargs) -> subprocess.CompletedProcess:
        return subprocess.run(
            ["ssh", self.host] + cmd,
            capture_output=True,
            text=True,
            **kwargs,
        )

    def read_file(self, path: str) -> str | None:
        result = self.run(["cat", path])
        if result.returncode == 0:
            return result.stdout
        return None

    def list_files(self, path: str) -> list[str]:
        result = self.run(["find", path, "-type", "f"])
        if result.returncode != 0:
            return []
        files = []
        for line in result.stdout.strip().splitlines():
            if line.startswith(path):
                rel = line[len(path) :].lstrip("/")
                if rel:
                    files.append(rel)
        return sorted(files)

    def deploy(
        self,
        staging_dir: Path,
        target_dir: str,
        sync_paths: list[str],
    ) -> list[str]:
        deployed = []

        # Ensure target dir exists
        self.run(["mkdir", "-p", target_dir])

        for sp in sync_paths:
            src = staging_dir / sp
            if not src.exists():
                continue

            remote_path = f"{self.host}:{target_dir}/{sp}"

            if src.is_dir():
                # rsync directory contents
                # Ensure parent dir exists on remote
                parent = f"{target_dir}/{sp}"
                self.run(["mkdir", "-p", parent])
                subprocess.run(
                    [
                        "rsync",
                        "-avz",
                        "--delete",
                        "-e",
                        "ssh",
                        f"{src}/",
                        f"{remote_path}/",
                    ],
                    check=True,
                    capture_output=True,
                )
            else:
                # rsync single file
                parent_dir = str(Path(f"{target_dir}/{sp}").parent)
                self.run(["mkdir", "-p", parent_dir])
                subprocess.run(
                    [
                        "rsync",
                        "-avz",
                        "-e",
                        "ssh",
                        str(src),
                        remote_path,
                    ],
                    check=True,
                    capture_output=True,
                )
            deployed.append(sp)

        return deployed

    def clean(self, target_dir: str, sync_paths: list[str]) -> None:
        for sp in sync_paths:
            self.run(["rm", "-rf", f"{target_dir}/{sp}"])

    @property
    def display_name(self) -> str:
        return f"ssh {self.host}"

File: src/adapters/ssh.ts
import * as fs from "fs";
import * as path from "path";
import { spawnSync } from "child_process";
import { Environment, type RunResult } from "./base.js";

export class SshEnvironment extends Environment {
  private host: string;

  constructor(host: string) {
    super();
    if (/^-/.test(host) || /\0/.test(host)) {
      throw new Error(`Invalid SSH host: ${host}`);
    }
    this.host = host;
  }

  isAvailable(): boolean {
    const result = spawnSync(
      "ssh",
      ["-o", "ConnectTimeout=5", "--", this.host, "true"],
      { encoding: "utf-8" },
    );
    return result.status === 0;
  }

  run(cmd: string[]): RunResult {
    const result = spawnSync("ssh", ["--", this.host, ...cmd], { encoding: "utf-8" });
    return {
      stdout: result.stdout ?? "",
      stderr: result.stderr ?? "",
      exitCode: result.status ?? 1,
    };
  }

  readFile(filePath: string): string | null {
    const result = this.run(["cat", filePath]);
    if (result.exitCode === 0) {
      return result.stdout;
    }
    return null;
  }

  listFiles(dirPath: string): string[] {
    const result = this.run(["find", dirPath, "-type", "f"]);
    if (result.exitCode !== 0) return [];
    const files: string[] = [];
    for (const line of result.stdout.trim().split("\n")) {
      if (line.startsWith(dirPath)) {
        const rel = line.slice(dirPath.length).replace(/^\//, "");
        if (rel) files.push(rel);
      }
    }
    return files.sort();
  }

  deploy(stagingDir: string, targetDir: string, syncPaths: string[]): string[] {
    const deployed: string[] = [];

    // Ensure target dir exists
    this.run(["mkdir", "-p", targetDir]);

    for (const sp of syncPaths) {
      const src = path.join(stagingDir, sp);
      // Check if source exists using node fs (staging is local)
      try {
        fs.statSync(src);
      } catch {
        continue;
      }

      const remotePath = `${this.host}:${targetDir}/${sp}`;
      const srcStat = fs.statSync(src);

      if (srcStat.isDirectory()) {
        const parentDir = `${targetDir}/${sp}`;
        this.run(["mkdir", "-p", parentDir]);
        spawnSync(
          "rsync",
          ["-avz", "--delete", "-e", "ssh", `${src}/`, `${remotePath}/`],
          { encoding: "utf-8" },
        );
      } else {
        const parentDir = path.dirname(`${targetDir}/${sp}`);
        this.run(["mkdir", "-p", parentDir]);
        spawnSync(
          "rsync",
          ["-avz", "-e", "ssh", src, remotePath],
          { encoding: "utf-8" },
        );
      }
      deployed.push(sp);
    }

    return deployed;
  }

  clean(targetDir: string, syncPaths: string[]): void {
    for (const sp of syncPaths) {
      this.run(["rm", "-rf", `${targetDir}/${sp}`]);
    }
  }

  get displayName(): string {
    return `ssh ${this.host}`;
  }
}
